improve CUDACachingAllocator lock contention

zyan0 opened the pull request:
Summary: NativeCachingAllocator has a global lock which shows lock contention with one process using multiple GPUs. The lock is required to lookup Block from pointer. We can make the lock more fine grain to reduce the lock contention.

Test Plan: existing unittests, verified on prod models using eight GPUs showing double digits improvements

Differential Revision: D52493091

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -2837,14 +2837,38 @@ void local_raw_delete(void* ptr);
-  std::mutex mutex;
+  static constexpr size_t kNumMutexShard = 1031;
+
+  // TODO: change to hardware_destructive_interference_size when C++17 is available.
```
PyTorch is fully c++17 compatible. Feel free to use it!

zyan0 replied:
thanks! It's actually not available in gcc/clang let me update the comment.

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -2837,14 +2837,38 @@ void local_raw_delete(void* ptr);
+  std::array<ska::flat_hash_map<void*, Block*>, kNumMutexShard>
+      allocated_blocks;
+
+  static constexpr uint64_t twang_mix64(uint64_t key) noexcept {
```
Is there a specific reason use this? Any builtin hash we could use here?

zyan0 replied:
the hash of an integer return the integer itself which is not helpful to balance the load

albanD replied:
I wasn't necessarily talking about std::hash but in dependencies we already have or even something from c10/util/hash.h ?

Thinking back at this, why isn't just doing `ptr % kNumMutexShard` would be unbalanced or maybe  `ptr >> 2 % kNumMutexShard` ? I guess our allocation lowest significant bits are pretty random.

houseroad replied:
seems c10::hash fallback to std::hash? https://codebrowser.bddppq.com/pytorch/pytorch/c10/util/hash.h.html#258

zyan0 replied:
> Thinking back at this, why isn't just doing `ptr % kNumMutexShard` would be unbalanced or maybe `ptr >> 2 % kNumMutexShard` ? I guess our allocation lowest significant bits are pretty random.

I think we can config alloc size to round up to a certain size. In that case, the lowest significant bits could be quite similar.

albanD replied:
Ok,
Would the sha1 implementation from c10/util/hash.h be fast enough for your use case here?
If not, adding a twang mix as a new hash sounds ok but I think it would be best to add it in next to the sha1 so that we can reuse it next time we need it.

zyan0 replied:
sounds good! sha1 looks slow for its std::string interface. let me move the function to hash.h.

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -2837,14 +2837,38 @@ void local_raw_delete(void* ptr);
+    std::mutex m;
+  };
+
+  std::array<AlignedMutex, kNumMutexShard> mutex;
```
Please add a block of comments explaining what we do and why.
Shard allocation region to have independent mutexes to reduce contention.

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -2837,14 +2837,38 @@ void local_raw_delete(void* ptr);
 class NativeCachingAllocator : public CUDAAllocator {
  private:
-  std::mutex mutex;
+  static constexpr size_t kNumMutexShard = 1031;
```
Any study on why this value? Any minimum level of concurrency we expect to smoothly support?

houseroad replied:
Same question, wondering how we come up with this magic number?

zyan0 replied:
I think it's possible to reduce this number without affecting lock contention, it's based on e2e tests so not super accurate and other workloads might be different

houseroad replied:
yeah, wondering if we reduce it to 63, how the perf looks like? I feel 1031 may be an overkill?

albanD replied:
Any concern about this array not fitting in a cache line or something similar that would push us to make it smaller?
If not, then I guess the justification is that we need whatever is "big enough" and, empirically, this is enough. But users can increase it if they see contention issues still happening.

zyan0 replied:
ok let me kept it in 63?

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -2836,14 +2838,28 @@ void local_raw_delete(void* ptr);
  private:
-  std::mutex mutex;
+  // Shard allocation region to have independent mutexes to reduce contention.
+  static constexpr size_t kNumMutexShard = 67;
```
You meant 63? :D

zyan0 replied:
63 is not a prime number so I found (using google lol) the nearest prime number which is 67.

albanD commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -28,6 +29,7 @@
 #include <iostream>
 #include <memory>
 #include <mutex>
+#include <new>
```
Since you don't use hardware_destructive_interference_size you don't this one right?

albanD reviewed (approved):
Thanks for updating the code.
Please do the two small nits before landing.

zyan0 commented:
@pytorchbot merge

Commit:
improve CUDACachingAllocator lock contention (#118550)

Summary:

NativeCachingAllocator has a global lock which shows lock contention with one process using multiple GPUs. The lock is required to lookup Block from pointer. We can make the lock more fine grain to reduce the lock contention.

bypass-github-pytorch-ci-checks

Test Plan: existing unittests, verified on prod models using eight GPUs showing double digits improvements

Reviewed By: albanD

Differential Revision: D52493091
